### PR TITLE
Use correct range for error in escape [ci: last-only]

### LINKED ISSUE #12981: Position failure under Bloop with small source file
## Questions are not bug reports

"What the -- ??"

## Reproduction steps

Scala version: 2.13.13

```scala
class Escapes { s"""aaa \s bbb""" }
```

## Problem

Correctly errors with
```
scala-cli compile --scala 2.13.13 --verbose --server=false short.scala
```
but crashes under Bloop.

Observed at https://github.com/scala/scala3/issues/20110

```
➜  bin ./scala-cli --version
Scala CLI version: 1.2.1
Scala version (default): 3.4.1
➜  bin cd
➜  ~ cd snips
➜  snips scala-cli compile --scala 2.13.13 short.scala
Downloading compilation server 1.5.16-sc-1
Starting compilation server
Compiling project (Scala 2.13.13, JVM (21))
Error compiling project (Scala 2.13.13, JVM (21))
Error: Unexpected error when compiling snips_e8caf7c7ca-2d67b5d8d6: java.lang.IndexOutOfBoundsException: 37
        at scala.reflect.internal.util.BatchSourceFile.offsetToLine(SourceFile.scala:213)
        at scala.tools.xsbt.DelegatingReporter$.makePosition$1(DelegatingReporter.scala:137)
        at scala.tools.xsbt.DelegatingReporter$.convert(DelegatingReporter.scala:190)
        at scala.tools.xsbt.DelegatingReporter.doReport(DelegatingReporter.scala:225)
        at scala.reflect.internal.Reporter.filteredInfo(Reporting.scala:126)
        at scala.reflect.internal.Reporter.error(Reporting.scala:121)
        at scala.tools.nsc.Reporting$PerRunReporting.error(Reporting.scala:377)
        at scala.tools.nsc.typechecker.Contexts$ImmediateReporter.error(Contexts.scala:1875)
        at scala.tools.nsc.typechecker.Contexts$ContextReporter.errorAndDumpIfDebug(Contexts.scala:1752)
        at scala.tools.nsc.typechecker.Contexts$Context.error(Contexts.scala:824)
        at scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen$.macroExpansionError(ContextErrors.scala:943)
        at scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen$.MacroGeneratedAbort(ContextErrors.scala:969)
        at scala.tools.nsc.typechecker.Macros.macroExpandWithRuntime(Macros.scala:817)
        at scala.tools.nsc.typechecker.Macros.macroExpandWithRuntime$(Macros.scala:772)
        at scala.tools.nsc.Global$$anon$6.macroExpandWithRuntime(Global.scala:512)
        at scala.tools.nsc.typechecker.Macros$MacroExpander.$anonfun$expand$2(Macros.scala:601)
        at scala.tools.nsc.Global.withInfoLevel(Global.scala:255)
        at scala.tools.nsc.typechecker.Macros$MacroExpander.expand(Macros.scala:594)
        at scala.tools.nsc.typechecker.Macros$MacroExpander.apply(Macros.scala:556)
        at scala.tools.nsc.typechecker.Macros.standardMacroExpand(Macros.scala:757)
        at scala.tools.nsc.typechecker.Macros.standardMacroExpand$(Macros.scala:755)
        at scala.tools.nsc.Global$$anon$6.standardMacroExpand(Global.scala:512)
        at scala.tools.nsc.typechecker.AnalyzerPlugins$$anon$11.default(AnalyzerPlugins.scala:474)
        at scala.tools.nsc.typechecker.AnalyzerPlugins$$anon$11.default(AnalyzerPlugins.scala:471)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.invoke(AnalyzerPlugins.scala:428)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsMacroExpand(AnalyzerPlugins.scala:471)
        at scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsMacroExpand$(AnalyzerPlugins.scala:471)
        at scala.tools.nsc.Global$$anon$6.pluginsMacroExpand(Global.scala:512)
        at scala.tools.nsc.typechecker.Macros.macroExpand(Macros.scala:746)
        at scala.tools.nsc.typechecker.Macros.macroExpand$(Macros.scala:739)
        at scala.tools.nsc.Global$$anon$6.macroExpand(Global.scala:512)
        at scala.tools.nsc.typechecker.Typers$Typer.vanillaAdapt$1(Typers.scala:1281)
        at scala.tools.nsc.typechecker.Typers$Typer.adapt(Typers.scala:1352)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:6264)
        at scala.tools.nsc.typechecker.Typers$Typer.typedStat$1(Typers.scala:6327)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$9(Typers.scala:3529)
        at scala.tools.nsc.typechecker.Typers$Typer.typedStats(Typers.scala:3529)
        at scala.tools.nsc.typechecker.Typers$Typer.typedTemplate(Typers.scala:2134)
        at scala.tools.nsc.typechecker.Typers$Typer.typedClassDef(Typers.scala:1972)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:6156)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:6249)
        at scala.tools.nsc.typechecker.Typers$Typer.typedStat$1(Typers.scala:6327)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$9(Typers.scala:3529)
        at scala.tools.nsc.typechecker.Typers$Typer.typedStats(Typers.scala:3529)
        at scala.tools.nsc.typechecker.Typers$Typer.typedPackageDef$1(Typers.scala:5836)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:6159)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:6249)
        at scala.tools.nsc.typechecker.Analyzer$typerFactory$TyperPhase.apply(Analyzer.scala:125)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:481)
        at scala.tools.nsc.typechecker.Analyzer$typerFactory$TyperPhase.run(Analyzer.scala:112)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1549)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1533)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1525)
        at scala.tools.nsc.Global$Run.compileFiles(Global.scala:1638)
        at scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:176)
        at scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:139)
        at scala.tools.xsbt.CompilerBridge.run(CompilerBridge.scala:43)
        at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.compileSources$1(BloopHighLevelCompiler.scala:133)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$7(BloopHighLevelCompiler.scala:159)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$1(BloopHighLevelCompiler.scala:71)
        at bloop.tracing.NoopTracer$.trace(BraveTracer.scala:53)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.timed$1(BloopHighLevelCompiler.scala:70)
        at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$6(BloopHighLevelCompiler.scala:159)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at monix.eval.internal.TaskRunLoop$.startFull(TaskRunLoop.scala:81)
        at monix.eval.internal.TaskRestartCallback.syncOnSuccess(TaskRestartCallback.scala:101)
        at monix.eval.internal.TaskRestartCallback.onSuccess(TaskRestartCallback.scala:74)
        at monix.eval.internal.TaskExecuteOn$AsyncRegister$$anon$1.run(TaskExecuteOn.scala:71)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

Compilation failed
```

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12981

This was broken since 2.13.0 `FastStringInterpolator`. The bad position is just a bad caret for ordinary reporting, but causes downstream errors for other tooling.

### CODE REVIEW

**Comment by som-snytt** [comment]
Apparently, the invariant, that point is in range, is not enforced.

I expected start <= point < end, but "zero extent" ranges are typical, also in dotty spans.

Splitting this commit into the pos fix for interpolator (plus gratuitous refactoring), then the tweak that you can request a spanning pos to include your default.point; otherwise a default.point out of range is ignored.

The tweaks improve point positioning, but as noted on the ancient PR for NamesDefault, it's difficult to code defensively because it's easy to accidentally derive a range via `offsetpos.withStart`. This PR is not ambitious because of likely interactions with tooling.

For example, `NamePos` gives the wrong position to a `var x` under `-Yrangepos:false`. Perhaps the option to turn off range positioning should be deprecated.

**Comment by som-snytt** [comment]
At any phase.

```
[info] [info] welcome to sbt 1.9.9 (Oracle Corporation Java 1.8.0_332)
[info] [info] loading project definition from /tmp/sbt_612ad334/project
[info] [info] loading settings for project sbt_612ad334 from common.sbt ...
[info] [info] set current project to sbt_612ad334 (in build file:/tmp/sbt_612ad334/)
[info] [info] compiling 2 Scala sources to /tmp/sbt_612ad334/target/classes ...
[info] [error] ## Exception when compiling 2 sources to /tmp/sbt_612ad334/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/collection/Seq$
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.Collection_SeqModule$lzycompute(Definitions.scala:507)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.Collection_SeqModule(Definitions.scala:507)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass$RunDefinitions.Seq_apply$lzycompute(Definitions.scala:1735)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass$RunDefinitions.Seq_apply(Definitions.scala:1733)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass$RunDefinitions.isSeqApply(Definitions.scala:1762)
[info] [error] scala.tools.nsc.transform.CleanUp$CleanUpTransformer.transformApply(CleanUp.scala:558)
[info] [error] scala.tools.nsc.transform.CleanUp$CleanUpTransformer.transform(CleanUp.scala:726)
[info] [error] scala.tools.nsc.transform.CleanUp$CleanUpTransformer.transform(CleanUp.scala:37)

```

File: test/files/neg/t12155.check [hunk 1]
```diff
@@ -1,4 +1,4 @@
 t12155.scala:6: error: class Node in class C1 cannot be accessed as a member of C1[A,_$1] from package <empty>
 class D[A, C](val x: C1[A, _]#Node[C]) {
-                              ^
+                     ^
```
Comment (som-snytt) [inline test/files/neg/t12155.check]: not an obvious improvement

**Comment by som-snytt** [comment]
Maybe it's time to follow Scala 3 and underline the range instead of the finicky point.

**Comment by som-snytt** [comment]
The failure on the first commit was weird.
```
[info] [info] compiling 1 Scala source to /tmp/sbt_8d42cfbb/target/classes ...
[info] [error] ## Exception when compiling 1 sources to /tmp/sbt_8d42cfbb/target/classes
[info] [error] java.util.ServiceConfigurationError: xsbti.compile.CompilerInterface2: jar:file:/home/jenkins/.ivy2/local/org.scala-lang/scala2-sbt-bridge/2.13.15-bin-SNAPSHOT/jars/scala2-sbt-bridge.jar!/META-INF/services/xsbti.compile.CompilerInterface2:1: Illegal provider-class name: 6scala.tools.xsbt.CompilerBridge
```

The failure on the last commit was the new check file from https://github.com/scala/scala/pull/10792

The checks were green before rebasing.

**Comment by lrytz** [comment]
Looks all good to me, thank you!

The change to `WrappingPosAccumulator` is a bit hard to asses. Did you maybe add some assertions to find call sites of `wrappingPos` / code examples where the `default.point` falls outside the wrapping range?

**Comment by SethTisue** [comment]
needs rebase

**Comment by som-snytt** [comment]
I intend to improve the code comment on widening pos. I believe I did start by asserting that point was in range.

**Comment by som-snytt** [comment]
I haven't looked at this again yet, but "For warnings" fails the test for positions of
```
val x, y = 42
```
where x overlaps with y, so x is focused on the x.

I didn't touch positions in that PR, but I see this touches mkClosure. Retronym's fix that induces focus was to supply the position of the rhs `42`. Maybe my rhs is no longer a range, so x no longer overlaps, so x is not focused (on my other PR).

It's by a pernicious coincidence that I must refresh my memory on positions, so it might be opportune to refresh this PR.

Edit: morning after, it was just a cut/paste error.

**Comment by som-snytt** [comment]
rebased on "for warnings".

**Comment by lrytz** [comment]
let me know when it's ready for another 👀

**Comment by som-snytt** [comment]
I was just looking for a new adventure in rebasing. Maybe I can figure out what to cherry-pick.

**Comment by som-snytt** [comment]
Not sure about the history of the semantics of position operations, but I see these tweaks are coping with `withStart` always producing a range but `withPoint` returning either a range or moving an offset position.

Symbolic ops seem to suggest notation, `x ^| y` for "which point do you prefer", etc, but the promise of convenience is not realized. Scala 2 is burdened with offset positions. Imagine if in 2014 the effort was to replace positions with spans on scala 2.

**Comment by som-snytt** [comment]
I rebased a previous rebase which left two commits, then adjusted some tests and also a few positions.

If a tree has the wrong position and there's no one there to hear it -- ?

**Comment by lrytz** [comment]
LGTM after squeaking 🐭

**Comment by som-snytt** [comment]
Squashed to the fix plus the tweak, with the suspicion that CB may suggest reverting it.

Thanks @lrytz for the extra extra :eyes:. I guess the squeaky wheel gets the grease, after all.

**Comment by SethTisue** [comment]
> Squashed to the fix plus the tweak, with the suspicion that CB may suggest reverting it.

There is one CB failure, a test failure in Metals: https://scala-ci.typesafe.com/job/scala-2.13.x-jdk17-integrate-community-build/2379/artifact/logs/metals-build.log

I haven't had time yet to investigate whether it's truly a regression, or whether the test is just overly sensitive.

**Comment by lrytz** [comment]
it's a completion test

```
[metals] [error] ==> X tests.pc.CompletionArgSuite.contructor-param_2.13.16-bin-06d9be6  0.047s munit.ComparisonFailException: tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:622
[metals] [error] 621:
[metals] [error] 622:  check(
[metals] [error] 623:    "contructor-param",
[metals] [error] Obtained empty output!
[metals] [error] => Expected:
[metals] [error] xxx = : Int
[metals] [error]     at munit.Assertions.failComparison(Assertions.scala:278)
```

https://github.com/scalameta/metals/blob/v1.4.1/tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala#L622-L632

```scala
  check(
    "contructor-param",
    """|class Foo (xxx: Int)
       |
       |object Main {
       |  val foo = new Foo(x@@)
       |}
       |""".stripMargin,
    """|xxx = : Int
       |""".stripMargin
  )
```

Maybe it's not this PR, but https://github.com/scala/scala/pull/10868 that is relevant?

**Comment by SethTisue** [comment]
hmm.... it's possible I got confused when bisecting — there was a confounding scalameta failure which made it impossible to test all the commits

**Comment by som-snytt** [comment]
Confound it!

**Comment by SethTisue** [comment]
It appears to be a real regression and not just a sensitive test.

I did the bisect again, this time without involving dbuild (by locally publishing scalameta as needed), and it still points to this PR and this PR only:

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
c014d322878218ff53340bf8c54d6de2d874a0b9
d856d1510ce8488914d0a4ae2bc9623bfb7de1c9
be51cc9b693ca8cfa061208635c600997dd8e348
```

I'll dig deeper early next week... unless the urge strikes this weekend...

**Comment by som-snytt** [comment]
I feel the urge more frequently these days.

Either positions are too fragile to touch or they are too fundamental to get wrong.

I can build a project locally if you tell me what failed.

**Comment by SethTisue** [comment]
@som-snytt let's pursue this further at https://github.com/scala/scala-dev/issues/891

**Pull request merged**